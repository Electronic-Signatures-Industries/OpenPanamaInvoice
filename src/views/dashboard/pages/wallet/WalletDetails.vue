<template>
  <v-expansion-panel-content>
    <v-list>
      <v-list-item two-line>
        <v-list-item-content>
          <v-list-item-title>DID</v-list-item-title>
          <v-list-item-subtitle>{{ this.didId }}</v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-list-item
        :two-line="this.balance === ''"
        :three-line="this.balance !== ''"
        >
        <v-list-item-content>
          <v-list-item-title>Ethereum Address</v-list-item-title>
          <v-list-item-subtitle>{{ this.currentAddress }}</v-list-item-subtitle>
          <v-list-item-subtitle v-if="this.balance !== ''">
            Balance: {{ this.balance }} {{ this.currencyName }}
          </v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-list-item two-line>
        <v-list-item-content>
          <v-list-item-title>Linked to Smart Card</v-list-item-title>
          <v-list-item-subtitle>{{ this.P11Name }}</v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-list-item two-line>
        <v-list-item-content>
          <v-list-item-title>Linked to P12</v-list-item-title>
          <v-list-item-subtitle>{{ this.P12Name }}</v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>

      <v-list-item two-line>
        <v-list-item-content>
          <v-list-item-title>Creation Date</v-list-item-title>
          <v-list-item-subtitle>{{ this.keystoreCreationDate }}</v-list-item-subtitle>
        </v-list-item-content>
      </v-list-item>
    </v-list>
  </v-expansion-panel-content>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from "vue-property-decorator";
import { KeystoreIndex } from "../shared/KeystoreIndex";
import Moment from "moment";
import { Contract } from "web3-eth-contract";
import Web3 from "web3";
import BN from "bn.js";

@Component({})
export default class WalletDetails extends Vue {
  @Prop()
  readonly didId: string;

  @Prop()
  readonly currentAddress: string;

  @Prop()
  readonly currentKeystore?: KeystoreIndex;

  @Prop()
  readonly erc20Contract?: Contract;

  balance = "";
  currencyName = "";

  get keystoreCreationDate(): string {
    if (!this.currentKeystore) return '';
    return Moment(this.currentKeystore.created).toISOString();
  }

  get P11Name() {
    return this?.currentKeystore?.linkedExternalKeystores?.pkcs11?.tokenIndex || "no";
  }

  get P12Name() {
    if (this?.currentKeystore?.linkedExternalKeystores?.pkcs12) {
      return (
        this.currentKeystore.linkedExternalKeystores.pkcs12?.name ||
        "No name found"
      );
    }

    return "no";
  }

  @Watch("erc20Contract")
  async setBalance(val: Contract) {
    if (!val || !this.currentAddress) {
      this.balance = "";
    }
    const [resultInWei, currencyName] = await Promise.all<BN, string>([
      val.methods.balanceOf(this.currentAddress).call(),
      val.methods.name().call()
    ]);
    this.balance = Web3.utils.fromWei(resultInWei);
    this.currencyName = currencyName || "tokens";
  }
}
</script>