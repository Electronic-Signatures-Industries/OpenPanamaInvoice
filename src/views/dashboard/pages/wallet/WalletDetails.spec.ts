import Vuetify from "vuetify";
import { createLocalVue, shallowMount, Wrapper } from "@vue/test-utils";
import WalletDetails from "./WalletDetails.vue";
import Web3 from "web3";

describe("WalletDetails.vue", () => {
  let component: Wrapper<WalletDetails>;

  beforeEach(() => {
    component = shallowMount(WalletDetails, {
      localVue: createLocalVue(),
      vuetify: new Vuetify(),
      propsData: {
        didId: "did:test",
        currentAddress: "0x000",
      },
    });
  });

  it("should match snapshot", async () => {
    expect(component).toMatchSnapshot();
  });

  it("should query correctly the Token Balance", async () => {
    // Mock the contract
    const balanceCall = jest.fn().mockResolvedValue(Web3.utils.toWei("20"));
    const nameCall = jest.fn().mockResolvedValue("Mock Coin");
    const mockContract = {
      methods: {
        balanceOf: jest.fn().mockReturnValue({ call: balanceCall }),
        name: jest.fn().mockReturnValue({ call: nameCall }),
      },
    };

    component.setProps({ erc20Contract: mockContract });
    await component.vm.$nextTick();
    await component.vm.$nextTick();
    await component.vm.$nextTick();

    expect(component.vm.$data.balance).toEqual("20");
    expect(component.vm.$data.currencyName).toEqual("Mock Coin");
  });
});
