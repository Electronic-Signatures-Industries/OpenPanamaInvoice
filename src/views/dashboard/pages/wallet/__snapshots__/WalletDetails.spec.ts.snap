// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletDetails.vue should match snapshot 1`] = `
<v-expansion-panel-content-stub>
  <v-list-stub tag="div">
    <v-list-item-stub activeclass="" tag="div" twoline="true">
      <v-list-item-content-stub>
        <v-list-item-title-stub>DID</v-list-item-title-stub>
        <v-list-item-subtitle-stub>did:test</v-list-item-subtitle-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
    <v-list-item-stub activeclass="" tag="div" twoline="true">
      <v-list-item-content-stub>
        <v-list-item-title-stub>Ethereum Address</v-list-item-title-stub>
        <v-list-item-subtitle-stub>0x000</v-list-item-subtitle-stub>
        <!---->
      </v-list-item-content-stub>
    </v-list-item-stub>
    <v-list-item-stub activeclass="" tag="div" twoline="true">
      <v-list-item-content-stub>
        <v-list-item-title-stub>Linked to Smart Card</v-list-item-title-stub>
        <v-list-item-subtitle-stub>no</v-list-item-subtitle-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
    <v-list-item-stub activeclass="" tag="div" twoline="true">
      <v-list-item-content-stub>
        <v-list-item-title-stub>Linked to P12</v-list-item-title-stub>
        <v-list-item-subtitle-stub>no</v-list-item-subtitle-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
    <v-list-item-stub activeclass="" tag="div" twoline="true">
      <v-list-item-content-stub>
        <v-list-item-title-stub>Creation Date</v-list-item-title-stub>
        <v-list-item-subtitle-stub></v-list-item-subtitle-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
  </v-list-stub>
</v-expansion-panel-content-stub>
`;
